core/WASM_core.md

```diff
@@ -0,0 +1,350 @@
+# Summary
+
+This proposal intends to outline the shape of the interfaces and implementation work necessary
+to make Core WASM-compilable.
+
+# Motivation
+
+A few problems a WASM Core Can solve, in no particular order:
+
+* Allow Go and Java to be ported to run on top of Core. *Potentially* without the downsides of CGo
+  and JNI. More on that later.
+* No longer any need to ship compiled Core binaries of various architectures with our existing SDKs.
+  This is both a nice plus for us, in terms of reducing CI complexity, but also a substantial size
+  reduction in our packages.
+* Workers in the browser
+* Hosted workers (also possible without, but with, there are some nice benefits)
+* A simpler way to manage callbacks into the lang layer, enabling things like user-defined metrics
+  exporters, lower-level interceptors, etc. This is certainly doable without WASM, but is probably
+  a bit easier (and, necessary) with it.
+* Auto-updates to core logic. This isn't detailed in this doc, but is a potentially very compelling
+  feature. Imagine a worker that can update core logic without needing to be restarted.
+
+
+# Detailed design
+
+The first and largest chunk of work involved in implementation is more or less pure grunt work. 
+I need to switch Core to compiling with the wasm32-unknown-unknown target, and then fix all the
+compiler errors.
+
+There are a few predictable problems we'll definitely have to deal with:
+
+The largest hiccup is the use of network (gRPC) calls. There is more than one way we could skin this
+cat. Luckily (intentionally), all of this is contained to the `temporal-client` crate.
+
+The next major source of side effects come from the tracing and metrics exporting facilities for
+OTel and Prometheus.
+
+Additionally, there is the code dealing with downloading and starting the CLI / Java test server.
+
+Finally, there may be some issues with other things that in native world use syscalls like atomics,
+locks, random number gen, etc. Luckily my experience so far is that those pretty much all just work
+without any, or fairly minor, changes.
+
+Ultimately we have two strategies available to us. We can create interfaces for Core to call back
+into lang to perform all these functions, or we can attempt to leverage WASI for this functionality.
+WASI is quite immature, but *in theory* it can support these use cases.
+
+## WASM Runtime
+
+Regardless of the strategy, we need to choose a WASM runtime which each Lang SDK will use, so I'll
```

Comment:
For TS, V8 can run WASM directly, there's no need for another WASM runtime.
For Go, I'm leaning towards wazero because of the CGo argument.
We'll need to assess what to do for Java but that will come in a later stage, by which the landscape may change.

Reply:
Can you show me the APIs you're thinking of using? This looks like it https://nodejs.dev/en/learn/nodejs-with-webassembly/

Which also mentions there's no WASI implementation. We may very well need at least basic WASI support for basic syscalls - so we may need to pick Wasmtime/Wasmer here anyway.

Reply:
Node has experimental WASI support: https://nodejs.org/api/wasi.html
Browser has this polyfill: https://wasi.dev/polyfill/
Using wasmtime or wasmer will require CGo, which is a no go (pun intended) IMHO.

I think it's worth exploring if we can avoid any syscalls other than memory allocation. We can easily handle memory allocation from the host.

Reply:
Yeah allocation is a given, and IME works everywhere without issue so far. The open questions for me are things like "Do locks and atomics work?" etc.

Reply:
I think it’s worth considering if we actually need multithreaded WASM core. Based on some initial digging it might not be straightforward to set up consistently across the different hosts.
V8 gives atomics that can be used to implement the other synchronization primitives and by leveraging multiple worker threads with shared memory you can build a multi threaded runtime.

Reply:
We probably don't really - and core is already largely not single threaded. The problem is actually expressing that in the type system means converting a bunch of things that are still `&self` currently to `&mut self`. Which means splitting the polling and completing halves of core into two, etc.

IE: Atomics and locks are needed not because of multi-threading, but because it is not encoded into the types that the _caller_ is single threaded, and even then you sometimes need synchronization between multiple mutable borrows on one thread. That never needs atomics (it could be RefCell instead), but a lot of the time it's easier to just use an atomic.

Regardless we'll just need to see what the support is like.

Reply:
What support do we need from our wasm runtime for atomics if it isn't actually being used and is just for the type system? It isn't super obvious to me if `wazero` supports threads/atomics

Reply:
If I'm using them, even if it's just for the type system, then they're needed and those calls are made.

I'm pretty sure wazero supports them, it seems fairly complete. Probably even the v8 one too, we just need to check. Typically they'll just be implemented with spinlocks or something if the actual underlying runtime isn't turning them into syscalls or the right CPU instruction.

Reply:
If we use a single threaded runtime all of the atomic operations are effectively noops apart from notify / signal.